src/Parallel/Algorithm.hpp

```diff
@@ -663,207 +746,74 @@ constexpr bool
+      if constexpr (Algorithm_detail::is_is_ready_callable_t<
+                        decltype(action),
+                        std::decay_t<decltype(check_local_box)>,
+                        tuples::tagged_tuple_from_typelist<inbox_tags_list>,
+                        Parallel::GlobalCache<metavariables>&,
+                        array_index>{}) {
+        return decltype(action)::is_ready(
+            check_local_box, std::as_const(inboxes_), *global_cache_,
+            std::as_const(array_index_));
+      } else {
+        return true;
+      }
+    };
 
     constexpr size_t phase_index =
         tmpl::index_of<phase_dependent_action_lists, PhaseDepActions>::value;
     using databox_phase_type = tmpl::at_c<databox_phase_types, phase_index>;
     using databox_types_this_phase = typename databox_phase_type::databox_types;
 
-    const auto display_databox_error = [this](const size_t line_number,
-                                              auto first_type,
-                                              auto... types) noexcept {
-      ERROR("The DataBox type being retrieved at algorithm step: "
-            << algorithm_step_ << " in phase " << phase_index
-            << " corresponding to action "
-            << pretty_type::get_name<this_action>() << " on line "
-            << line_number
-            << " is not the correct type but is of variant index "
-            << box_.which()
-            << ". If you are using Goto and Label actions then you are using "
-               "them incorrectly. \nValid DataBox Types: \n  "
-            << pretty_type::get_name<typename decltype(first_type)::type>()
-            << (MakeString{} << ",\n  "
-                << ...
-                << pretty_type::get_name<typename decltype(types)::type>())
-            << "\nVariant type:\n  " << pretty_type::get_name<variant_boxes>());
-    };
+    using box_possibilities = std::conditional_t<
+        iter == 0_st,
+        tmpl::list<
+            std::integral_constant<size_t, 0_st>,
+            std::integral_constant<
+                size_t, tmpl::size<databox_types_this_phase>::value - 1_st>>,
+        tmpl::list<std::integral_constant<size_t, iter>>>;
+    bool box_found = false;
+    tmpl::for_each<box_possibilities>([this, &box_found, &take_next_action,
+                                       &check_if_ready](
+                                          auto box_possibility_v) noexcept {
+      const size_t box_possibility = decltype(box_possibility_v)::type::value;
+      using this_databox =
+          tmpl::at_c<databox_types_this_phase, box_possibility>;
+      if (not box_found and
+          box_.which() ==
+              static_cast<int>(
+                  tmpl::index_of<variant_boxes, this_databox>::value)) {
+        box_found = true;
+        auto& box = boost::get<this_databox>(box_);
+        using local_this_action = tmpl::at_c<actions_list, iter>;
```

Comment:
Do you still need this?  I think it was for handling the special case for iter 0, which you are doing differently now.